`ultralytics 8.3.143` Add Profiling for Solutions

RizwanMunawar opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This update adds performance profiling to Ultralytics Solutions, allowing users to see how fast tracking and processing steps run, and improves logging for better transparency. 🚀⏱️

### 📊 Key Changes
- Added profiling to measure and report the speed of tracking/prediction and overall solution processing.
- Updated logs to show detailed timing information per frame when verbose mode is enabled.
- Included speed metrics in the results returned by solution classes.
- Minor code improvements and version bump to 8.3.143.

### 🎯 Purpose & Impact
- **Better Transparency:** Users can now see how long each step takes, making it easier to identify bottlenecks or performance issues.
- **Performance Monitoring:** Developers and users can track and optimize the speed of their workflows.
- **Improved Debugging:** Enhanced logs help with troubleshooting and understanding solution behavior.
- **No Breaking Changes:** All existing functionality remains, with added insights for those who want them.

Overall, this update makes it easier to monitor and optimize the performance of Ultralytics Solutions! 🚀📈

RizwanMunawar committed:
add ops.profiler

RizwanMunawar committed:
add ops.profiler

RizwanMunawar committed:
update

UltralyticsAssistant commented:
👋 Hello @RizwanMunawar, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help guide your contribution—an Ultralytics engineer will review and assist you soon.

To ensure a smooth review and integration, please check the following:

- ✅ **Define a Purpose**: Make sure the purpose of your feature or fix is clearly explained in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure commit messages are clear and follow project conventions.
- ✅ **Synchronize with Source**: Confirm your PR is up to date with the `main` branch of `ultralytics/ultralytics`. If not, update it by clicking the 'Update branch' button or running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Check that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If there are failures, please address them.
- ✅ **Update Documentation**: Update or add [documentation](https://docs.ultralytics.com/) for any new or changed features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes and confirm they all pass.
- ✅ **Sign the CLA**: If this is your first Ultralytics PR, please sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" on this PR.
- ✅ **Minimize Changes**: Only include the **minimum** necessary changes for your bug fix or feature. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, see our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have questions, feel free to leave a comment!

Thank you for contributing to Ultralytics! 🛠️✨

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'main' into sol-logger

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'sol-logger' of https://github.com/ultralytics/ultralytics into sol-logger

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'sol-logger' of https://github.com/ultralytics/ultralytics into sol-logger

RizwanMunawar committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'sol-logger' of https://github.com/ultralytics/ultralytics into sol-logger

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'sol-logger' of https://github.com/ultralytics/ultralytics into sol-logger

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar commented:
@glenn-jocher @Laughing-q Hi, this PR adds the profiling for Solutions. Please let me know what you think. Here's more information:

1. **`self.profilers[0]`**: This measures the time taken specifically for object tracking. It corresponds to: ```self.tracks = self.model.track(source=im0, persist=True, classes=self.classes, **self.track_add_args)```

2. **`self.profilers[1]`**: This covers the overall time spent inside the `process` method, which applies to all solution types (except semantic search and streamlit application, as these are browser-based applications). Everything executed within the `process` will be profiled here.
 
3. **`self.profilers[2]`**: This measures only the time taken for visualization using OpenCV (i.e., displaying the output image). If `show=False`, this value will always be zero. 

   - Since tracking and visualisation are part of the `process` method, we subtract their times to isolate the actual "solution" time in logs, i.e:

      ```python
      # Solution time = process - track - visualization
      f"{((self.profilers[1].dt - self.profilers[0].dt) - self.profilers[2].dt) * 1e3:.1f}ms solution, "
      ```

Additionally, I added a safeguard for `cv2.setMouseCallback` to prevent errors when `imshow` is disabled or unsupported (e.g., headless environments). Thanks

I just triggered the [Scheduled CI](https://github.com/ultralytics/ultralytics/actions/runs/15157525692) and [Docker Tests](https://github.com/ultralytics/ultralytics/actions/runs/15157540361), hoping everything will pass, and then we can merge this PR. Thanks

Laughing-q commented:
@RizwanMunawar Do we really want to profile the time of opencv visualization?

RizwanMunawar commented:
@Laughing-q yes, I notice, the visualisation itself takes a lot of time, i.e in some cases, visualisation is taking 5-10 ms per frame, so I think it will be better to add it. What do you think?

Laughing-q commented:
@RizwanMunawar wait I noticed it's not the plotting time, it's actually the imshow time.
https://github.com/ultralytics/ultralytics/blob/sol-logger/ultralytics/solutions/vision_eye.py#L65-L66
https://github.com/ultralytics/ultralytics/blob/b634c525f75a19de45aea53d60712d4cf4dce4c4/ultralytics/solutions/solutions.py#L194-L218
I don't think we need to profile this.

RizwanMunawar commented:
Yes, it's `cv2.imshow`, it's actually visualisation time, I guess not plotting time. Ok, no worries. Let me update the PR.

RizwanMunawar committed:
remove visualization logger

RizwanMunawar committed:
remove visualization logger

RizwanMunawar committed:
remove visualization logger

RizwanMunawar commented:
@Laughing-q ok, removed the visualisation profiling part, now only `track` and `solution` profiling we have, minimal changes :)

glenn-jocher committed:
Merge branch 'main' into sol-logger

Laughing-q commented on ultralytics/solutions/solutions.py:
```diff
@@ -209,9 +221,16 @@ def process(self, *args, **kwargs):
+                f"{self.profilers[0].dt * 1e3:.1f}ms track, "  # Object tracking time
+                # Solution time = process - track
+                f"{(self.profilers[1].dt - self.profilers[0].dt) * 1e3:.1f}ms solution, processing per image at shape "
+                f"(1, {getattr(self.model, 'ch', 3)}, {self.h}, {self.w})",
```
@RizwanMunawar I remember every solution returns `SolutionResults` no? and the `plot_im` stands for plotted image as an attribute there, can we get width and height from the `plot_im`? so we don't need to introduce these two `h` and `w` attributes.

RizwanMunawar replied:
Yes, I think this will work. I will adjust the PR.

Laughing-q commented on ultralytics/solutions/solutions.py:
```diff
@@ -209,9 +221,16 @@ def process(self, *args, **kwargs):
         if self.CFG["verbose"]:  # extract verbose value to display the output logs if True
-            LOGGER.info(f"🚀 Results: {result}")
+            LOGGER.info(
+                f"{self.profilers[0].dt * 1e3:.1f}ms track, "  # Object tracking time
```
@RizwanMunawar here's `track` in the description but for `object_crop` it's doing inference without track no?

RizwanMunawar replied:
@Laughing-q, yes, it's an exceptional case, as blur doesn't support the track method. We can add the check, but I don't think there will be a lot of difference in terms of processing time.

Laughing-q replied:
@RizwanMunawar oh I mean the `track` in the description, since some solutions do not use tracker at all like `blur` and `object_crop` and we are showing this `track` for these two as well I think.

Laughing-q commented on ultralytics/solutions/solutions.py:
```diff
@@ -136,7 +144,11 @@ def extract_tracks(self, im0):
+        with self.profilers[0]:
+            self.tracks = self.model.track(
+                source=im0, persist=True, classes=self.classes, verbose=False, **self.track_add_args
+            )
```
@RizwanMunawar probably we don't need to explicitly profile here? as we could get all the inference/preprocessing/postprocessing time directly from ultralytics predictor.
https://github.com/ultralytics/ultralytics/blob/93f54dcf9b1da5440ad5b3c043d9711d1994d712/ultralytics/engine/predictor.py#L335-L339

RizwanMunawar replied:
@Laughing-q, actually no, I guess it's not the same time solutions are using? There is an additional computation that will not be a part of this?

Laughing-q replied:
yeah I know, it's just kinda duplicate there, doing a overall profile while we have small pieces of profile inside predictor.

RizwanMunawar committed:
use results for image dimensions

RizwanMunawar commented:
@glenn-jocher Ok, here's the status of this PR, after detailed discussion with @Laughing-q:

- Initially, I was thinking it would be better to display the visualisation time, and I used 3 profilers, but Jing mentioned we can skip the visualisation part, so I adjusted it accordingly, and now we only have 2 profilers (1 for `track` and 1 for `solutions`).
- Initially, I extracted the image width and height for display, but `SolutionResults` also returns the processed image, so I adjusted the code to use its shape directly. It will avoid extra variables initialisation.

This PR is good to go from myside. Triggered the [Scheduled CI](https://github.com/ultralytics/ultralytics/actions/runs/15163915620) and [Docker Tests](https://github.com/ultralytics/ultralytics/actions/runs/15163902620), hoping everything will pass. Please let me know If any additional changes you think we should make? Thanks

glenn-jocher commented:
@RizwanMunawar I'm a little confused. Is the inference time not shown or is it included in one of the two shown?

Maybe for tracking it makes more sense to globally add a track time in the Results object alongside the rest of the speeds? Right now we have preprocessing, inference, and postprocessing speeds shown there.

glenn-jocher commented:
Speeds aren't primarily intended for showing in the console, they're intended for embedding in the result

RizwanMunawar committed:
Merge branch 'main' into sol-logger

RizwanMunawar committed:
Store speed results in SolutionResults

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented:
@glenn-jocher actually, in the existing implementation, we are displaying the mentioned info for solutions,  i.e, if we run the command `yolo solutions count`, we are getting the mentioned output in the console.
```
0: 384x640 17 persons, 12.2ms
Speed: 1.1ms preprocess, 12.2ms inference, 2.1ms postprocess per image at shape (1, 3, 384, 640)
```

You can see we have the preprocess, inference and postprocess speed, but nothing related to solutions. Since the solution core is the tracker, and the additional processing we do for each solution type. I created separate profilers to display the speed more precisely. You are right, we should add the profiling results in the `SolutionResults` object. I just updated the PR. The results of this PR with the command `yolo solutions count` are mentioned.

```
54: 360x640 3.8ms
Speed: 24.4ms track, 3.8ms solution per image at shape (1, 3, 360, 640)
```

And here's the `SolutionResults` object structure.

```json
{
  "in_count": 4,
  "total_tracks": 16,
  "classwise_count": {
    "person": {
      "IN": 4,
      "OUT": 0
    }
  },
  "speed": {
    "track": 31.90609999001026,
    "solution": 3.023400029633194
  }
}
```

We can also add the track profiling with the already existing profilers we have for `predictor`, but still solution time will not be covered. Please let me know what you think. Thanks

glenn-jocher commented:
@RizwanMunawar ok I see the problem, that the results object from the predictor is created before the solutions start running right? So that that time we can't add track and solution speeds?

RizwanMunawar commented:
@glenn-jocher yes exactly!

Normally, our solution structure is:
- predict or track (most of the solutions)
- additional processing for each solution

glenn-jocher committed:
Merge branch 'main' into sol-logger

glenn-jocher commented:
Ok got it, then I think this is good to merge

glenn-jocher committed:
Update __init__.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

UltralyticsAssistant commented:
🎉 Fantastic work, @RizwanMunawar, @glenn-jocher, and @Laughing-q! Merging this PR marks a significant step forward for Ultralytics Solutions. Your addition of performance profiling and enhanced logging empowers users with deeper insights and transparency.

As Thomas Edison once said, “There’s a way to do it better—find it.” Your commitment to continuous improvement truly embodies this spirit, making it easier for everyone to monitor, optimize, and trust their workflows.

Thank you for your dedication and impactful contributions—together, we’re building a smarter, faster future with Ultralytics! 🚀
